105 login page frontend (work in progress!)

natapokie opened the pull request:
looks like this on desktop,
![image](https://user-images.githubusercontent.com/86009011/173122379-f7819932-4c8d-4eab-b35d-ffedf9eb39cd.png)
![image](https://user-images.githubusercontent.com/86009011/173122451-3a6ba90b-409e-4e6f-a85d-271af745cb1a.png)


on mobile,
![image](https://user-images.githubusercontent.com/86009011/173121642-0a151274-3f9e-4f3e-b768-ec700e7adb82.png)

one problem is the footer, would it be okay if it doesn't show up on all pages?

jameskokoska commented on client/src/util/pages.jsx:
```diff
@@ -30,7 +31,7 @@ export const pages = {
     {
       label: 'Login',
-      component: <div />,
+      component: <PageLogin />,
```
Special isn't spread into the map of the navigation framework of the main app
Basically add this in App.jsx 
![image](https://user-images.githubusercontent.com/50821962/173151556-fc35007f-185e-4b03-aa66-34008d61488e.png)

natapokie replied:
ohh i see! thank you!

jameskokoska commented:
To remove the Footer on some pages you can do something like
![image](https://user-images.githubusercontent.com/50821962/173151869-6b4dc3a6-c7df-457e-a4f5-a362e0c9ad01.png)
![image](https://user-images.githubusercontent.com/50821962/173151847-2c5e9b52-6637-48de-9b42-739c9a5504c5.png)

jameskokoska commented on client/src/pages/Login/Login.jsx:
```diff
@@ -0,0 +1,62 @@
+    <div className="login-entire-page">
+      <div className="login-bg">
+        <div className="login-container">
+          <h2 className="login-title">Login</h2>
```
You should make this text larger

jameskokoska commented on client/src/pages/Login/Login.jsx:
```diff
@@ -0,0 +1,62 @@
+      <div className="login-bg">
+        <div className="login-container">
+          <h2 className="login-title">Login</h2>
+          <TextInput inputType={'text'} placeholder={'Username or Email'}></TextInput>
```
I can't seem to enter anything
Maybe the layers (images) are overlapping

jameskokoska replied:
It'll just be email
no need to call it Username or Email

jameskokoska commented on client/src/pages/Login/Login.jsx:
```diff
@@ -0,0 +1,62 @@
+
+          <div className="login-button-container">
+            <p className="forgot-message">Forgot Password?</p>
+            <Button label={'Log in'}></Button>
```
When the login button is pressed it should check the credentials and produce an error message
If you need help with this let me know, it requires state management
And to set the error if one is produced from the backend

Also put your functions in a separate file
(see https://github.com/UofT-Frosh-Orientation/orientation-website/pull/125 for e.g. of separate file functions)

jameskokoska commented on client/src/pages/Login/Login.jsx:
```diff
@@ -0,0 +1,62 @@
+          {/* {incorrectEntry ? <p className="login-error-message">Incorrect Username or Password, please try again!</p> : <></>} */}
+
+          <div className="login-button-container">
+            <p className="forgot-message">Forgot Password?</p>
```
If a user hits forget password
We should make a forgot password page or popup
Maybe when the popup component is done or you can make it so the user just needs to enter their username to get an email with a new password

Again requires state management see comment below

jameskokoska reviewed (changes requested):
Looks good! I like the background

natapokie commented on client/src/components/popup/PopupModal.jsx:
```diff
@@ -0,0 +1,58 @@
+    <div className="popup-background" style={{ width: `${bgWidth}`, height: `${bgHeight}` }}>
+      <div
+        className="popup-container"
+        style={{ width: `${containerWidth}`, top: `${containerTop}` }}
```
```
        style={{ width: `${containerWidth}`, top: `${containerTop}` }}
```
hi @jameskokoska quick question! i'm trying to turn the component i made for the password popup, into a "reusable" component, and I want to pass another set of parameters for mobile/tablet display but i don't know the syntax 🤔

jameskokoska replied:
so container width and container top would vary based on if it's mobile or not?
The best way to do this is to use CSS classes

natapokie replied:
yep, for the forgot password popup, i just wanted to make the width wider in mobile, so it's easier for people to enter emails... and also container top is kinda weird because the height of the login page with 150vh (in desktop) and then 100vh (in mobile), so i wanted to adjust the placement of the popup a little...😅

natapokie commented on client/src/pages/Login/Login.scss:
```diff
@@ -0,0 +1,374 @@
+  align-items: center;
+}
+
+// POPUP ANIMATION
```
@jameskokoska more questions... I also added a popup animation when you open the popup, but I don't know how to animate the closing part, if you have any suggestions that would be great!

jameskokoska replied:
Easiest way is to use transition
Transition will automatically reverse the animation when the style is removed

natapokie replied:
i might be doing transition wrong because it doesn't seem to be working... but I'll look at last year's code!

jameskokoska commented on client/src/components/input/TextInput/TextInput.jsx:
```diff
@@ -19,6 +19,12 @@ const TextInput = ({
   hasRestrictedInput,
   inputType,
 }) => {
+  useEffect(() => {
```
Merge with dev (to avoid conflicts, if you haven't already), I had to add this change there for registration

jameskokoska commented on client/src/pages/Login/Login.jsx:
```diff
@@ -0,0 +1,155 @@
+            const result = resetPassword(email);
+
+            if (result !== true) {
+              setEmailError(result);
```
After submitting a request to change password, the `Send` button should disappear and be replaced with a `close` popup button

natapokie replied:
sounds good! i'll use the errorBox, once it's merged to dev 👍

jameskokoska commented on client/src/components/popup/PopupModal.jsx:
```diff
@@ -0,0 +1,58 @@
+        className="popup-container"
+        style={{ width: `${containerWidth}`, top: `${containerTop}` }}
+      >
+        <img
```
The image should be larger, hit box small

jameskokoska commented on client/src/components/popup/PopupModal.jsx:
```diff
@@ -0,0 +1,58 @@
+import './PopupModal.scss';
+import XMark from '../../assets/misc/xmark-solid-white.svg';
+
+const PopupModal = ({
```
Tapping the background, should close the popup
You can use something like `useWrapperRef.jsx`
It was used in the `Dropdown` component

jameskokoska reviewed (changes requested):
Do you want to add an error message box at the bottom instead of in the password field
I made an errorbox component in https://github.com/UofT-Frosh-Orientation/orientation-website/pull/164

Example:
![image](https://user-images.githubusercontent.com/50821962/174338818-23af8062-8032-47b6-b2d8-11b4e60ddb1d.png)

For the popup: see last years, you can try to port over the code or use as reference
https://github.com/UofT-Frosh-Orientation/froshweek3.0/blob/dev/client/src/components/containers.js#L179

Video:

https://user-images.githubusercontent.com/50821962/174340355-4af04607-a899-423a-b627-43d1e009f58a.mp4

jameskokoska commented:
Looks good though, I like the animated dinos

jameskokoska commented:
@natapokie If you want the login to succeed, just return `true` from this function, instead of `result`
This promise just adds a delay in an async function, similar to requesting a response to a server and waiting until the credentials are valid/invalid
Here we wait `1000ms` to simulate awaiting a response from the server backend api
```js
export async function login(username, password) {
  console.log(username, password);
  /*eslint no-undef: 0*/
  let promise = new Promise((res, rej) => {
    setTimeout(() => res('An error occured. Please try again.'), 1000);
  });
  let result = await promise;
  return result; //return an error message string to be dispayed, if an error
}
```

Cali0707 commented on client/src/components/schedule/ScheduleHome/ScheduleHome.jsx:
```diff
@@ -0,0 +1,137 @@
+  const buttonList = Object.keys(data).map((item) => {
+    return { name: item };
+  });
+  console.log(closeAll);
```
```suggestion

```

Cali0707 commented on client/src/components/schedule/ScheduleHome/ScheduleHome.jsx:
```diff
@@ -0,0 +1,137 @@
+function getDaysSchedule() {
+  const scheduleData = data;
+  const days = [];
+  console.log(Object.keys(scheduleData));
```
```suggestion

```

Cali0707 commented on client/src/components/text/Accordion/SingleAccordion/SingleAccordion.jsx:
```diff
@@ -3,7 +3,7 @@ import PropTypes from 'prop-types';
 
-const SingleAccordion = ({ header, children, isOpen, setIsOpen }) => {
+const SingleAccordion = ({ header, children, isOpen, setIsOpen, canOpen, className }) => {
   const [Height, setHeight] = useState('0px');
   const [Rotate, setRotate] = useState('accord-icon');
```
```suggestion
  const [height, setHeight] = useState('0px');
  const [rotate, setRotate] = useState('accord-icon');
```

Cali0707 commented on client/src/components/text/Accordion/SingleAccordion/SingleAccordion.jsx:
```diff
@@ -15,16 +15,24 @@ const SingleAccordion = ({ header, children, isOpen, setIsOpen }) => {
+          <></>
+        )}
       </div>
       <div ref={content} style={{ maxHeight: `${Height}` }} className={'accord-content'}>
```
```suggestion
      <div ref={content} style={{ maxHeight: `${height}` }} className={'accord-content'}>
```

Cali0707 commented on client/src/components/text/Accordion/SingleAccordion/SingleAccordion.jsx:
```diff
@@ -15,16 +15,24 @@ const SingleAccordion = ({ header, children, isOpen, setIsOpen }) => {
+        {header}
+        {canOpen !== false ? (
+          <div style={{ marginLeft: 'auto' }} className={'accord-text'}>
+            <img src={faAngleDown} className={`${Rotate}`} alt="Accordion Button" width="15px" />
```
```suggestion
            <img src={faAngleDown} className={`${rotate}`} alt="Accordion Button" width="15px" />
```

Cali0707 commented on client/src/components/text/Accordion/SingleAccordion/SingleAccordion.scss:
```diff
@@ -2,9 +2,8 @@
 .accordion {
   padding: 0px 20px 0px 20px;
-  background-color: var(--purple-shades-light);
+  // background-color: var(--purple-shades-light);
```
```suggestion

```

Cali0707 commented on client/src/pages/Login/Login.jsx:
```diff
@@ -0,0 +1,167 @@
+  const [loginError, setLoginError] = useState('');
+  const [loginState, setLoginState] = useState('');
+
+  let username = '';
+  let password = '';
```
These should be state variables

Cali0707 commented on client/src/pages/Login/functions.jsx:
```diff
@@ -0,0 +1,19 @@
+export async function login(username, password) {
+  console.log(username, password);
+  /*eslint no-undef: 0*/
+  let promise = new Promise((res, rej) => {
```
@natapokie can you send me a list of all the places (like this) in this PR where we should be calling to an API but currently are not?

natapokie replied:
@Cali0707 the functions in this file, `login `and `resetPassword `need to be calling to the email api

Cali0707 commented on client/src/pages/Login/functions.jsx:
```diff
@@ -0,0 +1,19 @@
+export async function login(username, password) {
+  console.log(username, password);
```
We should not log password ever in production
```suggestion

```

Cali0707 commented on client/src/pages/Login/functions.jsx:
```diff
@@ -0,0 +1,19 @@
+}
+
+export function resetPassword(email) {
+  console.log(email);
```
```suggestion

```

jameskokoska commented:
General feedback - up to you if you want to change

Cursor shouldn't be pointer finger when hovering over the background
Content inside the popup doesn't scroll when it overflows (offscreen)
Popup gets clipped when content too large (not centered perfectly)
Padding around exit button should be larger
Popup is open for a split second when page loads - fix: should have display: none if it hasn't been opened once 
Password reset: The success message when sending an email should use the errorbox component and be reflective if success or error
The send button should disappear if sending a password reset email was successful
Popup animation seems a bit harsh when opening, should be toned down

natapokie commented:
most recent commit...  754a37b, fixed issues with the popup component!

![image](https://user-images.githubusercontent.com/86009011/175721453-01b6ef4f-774d-433a-af0b-32e0208b2688.png)

@jameskokoska if you can review the popup component when you have time, that would be great, thank you! I think I fixed all the problems with the popup... but let me know if I missed anything 👍
I rewrote the popup so there are a lot of comments, which I delete after I receive an "okay" from you !

jameskokoska commented:
Great job with the popup!
Some suggestions
- [ ] The close and open animation, feel free to make it a bit more smooth
- [ ] You can increase the max vertical height of the popup on mobile devices (obviously if the popup content isn't enough to fit though it would take the smaller size, but I think you already have that figured out)
- [ ] For the login page, make it run the `login` action when the user hit's enter in the password field (use the `onEnterKey` prop of TextInput)
- [ ] Same for the password reset send email TextInput, use the `onEnterKey` prop

natapokie commented:
in latest commits 9f671ce and eae9461 , added the reset password email checks...!

@Cali0707 the places that need to be connected to the email api are in Login/functions.jsx -- I'll leave comments in the code as well!
![image](https://user-images.githubusercontent.com/86009011/175757823-d4a9cba3-1f81-4913-aeb2-614ecb02b265.png)

reset password popup looks like this now...
![image](https://user-images.githubusercontent.com/86009011/175757666-71583e28-8e8d-4bc9-8edd-b5cc0a2d662d.png)

when async function returns false...
![image](https://user-images.githubusercontent.com/86009011/175757683-d1114df7-4127-403a-8396-248c0d9301f1.png)

when async function returns true...
![image](https://user-images.githubusercontent.com/86009011/175757699-0daa80a1-16dd-4726-b3f7-2c5e5f55c88d.png)

natapokie commented on client/src/pages/Login/functions.jsx:
```diff
@@ -0,0 +1,25 @@
+export async function resetPassword(email) {
+  console.log(email);
+  let promise = new Promise((res, rej) => {
+    setTimeout(() => res(''), 1000);
```
@Cali0707 i wasn't sure if you could return two responses... so I just displayed the message (i.e. email sent, email not found) on the front end based on the result (if it was true or false)!

natapokie commented on client/src/pages/Login/Login.jsx:
```diff
@@ -0,0 +1,241 @@
+  const [emailInput, setEmailInput] = useState(''); // email entered by user
+  const [buttonClick, setButtonClick] = useState(0);
+
+  async function resetButtonPress() {
```
```
  async function resetButtonPress() {
```

@Cali0707  this is the function that calls the async resetPassword function and changes states accordingly!

natapokie commented:
> * [ ]  The close and open animation, feel free to make it a bit more smooth

@jameskokoska I think I made the fade in animation more smooth, but the fadeout is a little tricky... 🤔 you can always see the text fade out last, it's kinda like all the children are fading out at different times (?), I was wondering if you have any suggestions I could try?

jameskokoska commented:
> in latest commits 9f671ce and eae9461 , added the reset password email checks...!
> 
> @Cali0707 the places that need to be connected to the email api are in Login/functions.jsx -- I'll leave comments in the code as well!


Create a backend ticket so that these need to be done after this PR gets submitted

jameskokoska commented:
> > * [ ]  The close and open animation, feel free to make it a bit more smooth
> 
> @jameskokoska I think I made the fade in animation more smooth, but the fadeout is a little tricky... 🤔 you can always see the text fade out last, it's kinda like all the children are fading out at different times (?), I was wondering if you have any suggestions I could try? 
> 
> 

I shall take a look

jameskokoska commented:
You can see what I changed here: https://github.com/UofT-Frosh-Orientation/orientation-website/pull/128/commits/edc04b4d56a3b83c88b3f9cf1587bc7492adb862

The main problem was this
![image](https://user-images.githubusercontent.com/50821962/175779122-44b5973c-73ff-4cd8-9a32-37441db7a4e5.png)
Would apply immediately
You can just use the animation of opacity to animate these 
Don't animate filter blur as it is very performance intensive, an opacity animation fakes this anyways

Also a made the exit button absolute

Cali0707 commented on client/src/components/popup/PopupModal.jsx:
```diff
@@ -0,0 +1,80 @@
+}) => {
+  const [hasBeenOpened, setHasBeenOpened] = useState(false);
+
+  if (hasBeenOpened === false && trigger === false) {
+    return <div></div>;
+  } else if (hasBeenOpened === false && trigger === true) {
+    setHasBeenOpened(true);
+  }
```
I would wrap this in a `useEffect` - React can sometimes be a little confusing otherwise with when it re-renders the components

Cali0707 reviewed (approved):
I left one piece of feedback, but otherwise this looks good! Great work @natapokie !!

natapokie commented:
> You can see what I changed here: [edc04b4](https://github.com/UofT-Frosh-Orientation/orientation-website/commit/edc04b4d56a3b83c88b3f9cf1587bc7492adb862)
> 
> The main problem was this ![image](https://user-images.githubusercontent.com/50821962/175779122-44b5973c-73ff-4cd8-9a32-37441db7a4e5.png) Would apply immediately You can just use the animation of opacity to animate these Don't animate filter blur as it is very performance intensive, an opacity animation fakes this anyways
> 
> Also I made the exit button absolute

Thanks for fixing it @jameskokoska, it looks so much better now !! I'll make sure to limit the animations to just opacity and transform! 👍👍

natapokie commented:
latest commit 13321d3, 
i added `useEffect`, the effect only executes for the first render of the popup, 
i took the `<div></div>` out of `useEffect`, so that when the page first renders, you won't see the popup for a split second.
@Cali0707 let me know if this is okay, it's a little different from your suggestion, but i think it almost works the same way !
